File: src/com/android/launcher3/applibrary/view/AppLibraryAdapter.java
package com.android.launcher3.applibrary.view;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.android.launcher3.R;
import com.android.launcher3.applibrary.model.AppCategoryGroup;
import com.android.launcher3.views.ActivityContext;

import java.util.ArrayList;
import java.util.List;

public class AppLibraryAdapter extends RecyclerView.Adapter<AppLibraryAdapter.ViewHolder> {

    private final ActivityContext mActivityContext;
    private final List<AppCategoryGroup> mGroups = new ArrayList<>();
    private AppCategoryCardView.OnCategoryExpandListener mExpandListener;

    public AppLibraryAdapter(ActivityContext activityContext) {
        mActivityContext = activityContext;
    }

    public void setOnCategoryExpandListener(AppCategoryCardView.OnCategoryExpandListener listener) {
        mExpandListener = listener;
    }

    public void setGroups(List<AppCategoryGroup> groups) {
        mGroups.clear();
        if (groups != null) {
            mGroups.addAll(groups);
        }
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        AppCategoryCardView view = (AppCategoryCardView) inflater.inflate(
                R.layout.app_library_category_card, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        AppCategoryGroup group = mGroups.get(position);
        holder.mCardView.setOnCategoryExpandListener(mExpandListener);
        holder.mCardView.bindGroup(group, mActivityContext);
    }

    @Override
    public int getItemCount() {
        return mGroups.size();
    }

    static class ViewHolder extends RecyclerView.ViewHolder {
        final AppCategoryCardView mCardView;

        ViewHolder(AppCategoryCardView cardView) {
            super(cardView);
            mCardView = cardView;
        }
    }
}

File: src/com/android/launcher3/applibrary/AppLibraryAdapter.kt
package com.android.launcher3.applibrary

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.android.launcher3.R
import com.android.launcher3.model.data.AppInfo

class AppLibraryAdapter : RecyclerView.Adapter<AppLibraryAdapter.CategoryViewHolder>() {
    
    private val categories = mutableListOf<Pair<String, List<AppInfo>>>()
    
    fun updateCategories(newCategories: Map<String, List<AppInfo>>) {
        categories.clear()
        categories.addAll(newCategories.toList())
        notifyDataSetChanged()
    }
    
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CategoryViewHolder {
        val view = LayoutInflater.from(parent.context)
            .inflate(R.layout.app_library_card, parent, false)
        return CategoryViewHolder(view as AppLibraryCardView)
    }
    
    override fun onBindViewHolder(holder: CategoryViewHolder, position: Int) {
        val (category, apps) = categories[position]
        holder.bind(category, apps)
    }
    
    override fun getItemCount(): Int = categories.size
    
    class CategoryViewHolder(private val cardView: AppLibraryCardView) : RecyclerView.ViewHolder(cardView) {
        fun bind(category: String, apps: List<AppInfo>) {
            cardView.setCategory(category, apps)
        }
    }
}
